# Calculate format=diff pkg(sys-kernel/calculate-sources)=>3.18.20,3.19&&merge(x11-drivers/ati-drivers)=>15.1,15.8
diff --git a/common/lib/modules/fglrx/build_mod/firegl_public.c b/common/lib/modules/fglrx/build_mod/firegl_public.c
index a8961ad..1287e80 100755
--- a/common/lib/modules/fglrx/build_mod/firegl_public.c
+++ b/common/lib/modules/fglrx/build_mod/firegl_public.c
@@ -4498,8 +4498,8 @@ static void kcl_mem_pat_setup (void *info)
 
     if (cpu_has_pge)
     {
-        cr4 = read_cr4();
-        write_cr4(cr4 & ~X86_CR4_PGE);
+        cr4 = __read_cr4();
+        __write_cr4(cr4 & ~X86_CR4_PGE);
     }
      __flush_tlb();
 
@@ -4512,7 +4512,7 @@ static void kcl_mem_pat_setup (void *info)
     write_cr0(cr0 & 0xbfffffff);
     if (cpu_has_pge)
     {
-        write_cr4(cr4);
+        __write_cr4(cr4);
     }
     local_irq_restore(flags);
 
@@ -4539,8 +4539,8 @@ static void kcl_mem_pat_restore (void *info)
 
     if (cpu_has_pge)
     {
-        cr4 = read_cr4();
-        write_cr4(cr4 & ~X86_CR4_PGE);
+        cr4 = __read_cr4();
+        __write_cr4(cr4 & ~X86_CR4_PGE);
     }
      __flush_tlb();
   
@@ -4552,7 +4552,7 @@ static void kcl_mem_pat_restore (void *info)
     write_cr0(cr0 & 0xbfffffff);
     if (cpu_has_pge)
     {
-        write_cr4(cr4);
+        __write_cr4(cr4);
     }
     local_irq_restore(flags);
 
diff --git a/common/lib/modules/fglrx/build_mod/kcl_acpi.c b/common/lib/modules/fglrx/build_mod/kcl_acpi.c
index 3a80984..679cfbf 100755
--- a/common/lib/modules/fglrx/build_mod/kcl_acpi.c
+++ b/common/lib/modules/fglrx/build_mod/kcl_acpi.c
@@ -861,7 +861,13 @@ void ATI_API_CALL KCL_ACPI_No_Hotplug(void* dev)
 #elif LINUX_VERSION_CODE >= KERNEL_VERSION(3,17,0)
     if(pdev)
     {
-       pci_ignore_hotplug(pdev);
+       //pci_ignore_hotplug(pdev);
+       struct pci_dev *bridge = pdev->bus->self;
+
+       pdev->ignore_hotplug = 1;
+       /* Propagate the "ignore hotplug" setting to the parent bridge. */
+       if (bridge)
+           bridge->ignore_hotplug = 1;
     }
 #endif
 }
